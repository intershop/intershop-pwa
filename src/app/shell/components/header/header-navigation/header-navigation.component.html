<ul class="navbar-nav main-navigation-list">
  <li class="home hidden-xs">
    <a [routerLink]="['/home']" title="home"><span class="glyphicon glyphicon-home"></span></a>
  </li>
  <li *ngFor="let category of categories" class="dropdown" [ngClass]="{'open' : category.isOpenedToggle}">
    <a [routerLink]="[categoriesService.generateCategoryRoute(category)]" [attr.data-testing-id]="category.id + '-' + link" [style.width]="{'100%': category.hasOnlineSubCategories}">{{category.name}}</a>
    <a *ngIf="category.subCategories" (click)="category.isOpenedToggle = !category.isOpenedToggle" class="dropdown-toggle visible-xs">
      <span class="glyphicon" [ngClass]="category.isOpenedToggle ? 'glyphicon-minus' : 'glyphicon-plus'"></span>
    </a>
    <is-subcategory-navigation [category]="category" [subCategoriesDepth]="1"></is-subcategory-navigation>
  </li>
</ul>
