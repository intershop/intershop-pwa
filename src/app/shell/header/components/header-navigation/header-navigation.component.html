<ul class="navbar-nav main-navigation-list d-md-none" [ngClass]="{ accordion: view !== 'full' }">
  <ng-container *ngTemplateOutlet="categoryTemplate"></ng-container>
</ul>
<ul class="navbar-nav main-navigation-list d-none d-md-block" [ngClass]="{ accordion: view === 'small' }">
  <ng-container *ngTemplateOutlet="categoryTemplate"></ng-container>
</ul>
<!--- Category Template -->
<ng-template #categoryTemplate>
  <li
    *ngFor="let category of categories"
    #submenu
    class="dropdown"
    [ngClass]="{ open: isOpened(category) }"
    (mouseover)="subMenuShow(submenu)"
    (mouseout)="subMenuHide(submenu)"
    (click)="subMenuHide(submenu)"
  >
    <a
      [routerLink]="['/category', category.uniqueId]"
      [attr.data-testing-id]="category.uniqueId + '-link'"
      [ngStyle]="{ width: !category.hasChildren() ? '100%' : '' }"
    >
      {{ category.name }}
    </a>
    <a
      *ngIf="category.hasChildren()"
      class="dropdown-toggle"
      [ngClass]="{ 'd-none': view === 'full', 'd-md-none': view === 'auto' }"
      (click)="toggleOpen(category)"
    >
      <fa-icon *ngIf="isOpened(category); else closed" [icon]="['fas', 'minus']"></fa-icon>
      <ng-template #closed> <fa-icon [icon]="['fas', 'plus']"></fa-icon> </ng-template>
    </a>
    <ish-sub-category-navigation
      [category]="category"
      [subCategoriesDepth]="1"
      [view]="view"
    ></ish-sub-category-navigation>
  </li>
</ng-template>
