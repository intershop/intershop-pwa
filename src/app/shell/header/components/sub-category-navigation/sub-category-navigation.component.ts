import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';

import { MAIN_NAVIGATION_MAX_SUB_CATEGORIES_DEPTH } from 'ish-core/configurations/injection-keys';
import { CategoryView } from 'ish-core/models/category-view/category-view.model';
import { Category } from 'ish-core/models/category/category.model';

/**
 * The Sub Category Navigation Component displays second level category navigation.
 *
 * @example
 * <ish-sub-category-navigation
 *   [category]="category"
 *   [subCategoriesDepth]="1"
 * >
 * </ish-sub-category-navigation>
 */
@Component({
  selector: 'ish-sub-category-navigation',
  templateUrl: './sub-category-navigation.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SubCategoryNavigationComponent {
  @Input() view = 'auto';
  @Input() category: CategoryView;
  @Input() subCategoriesDepth: number;

  openedCategories = [];

  constructor(@Inject(MAIN_NAVIGATION_MAX_SUB_CATEGORIES_DEPTH) public mainNavigationMaxSubCategoriesDepth: number) {}

  /**
   * Indicate if specific category is expanded.
   * @param category The category item.
   */
  isOpened(category: Category): boolean {
    return this.openedCategories.includes(category.uniqueId);
  }

  /**
   * Toggle category open state.
   * @param category The category item.
   */
  toggleOpen(category: Category) {
    const index = this.openedCategories.findIndex(id => id === category.uniqueId);
    index > -1 ? this.openedCategories.splice(index, 1) : this.openedCategories.push(category.uniqueId);
  }
}
