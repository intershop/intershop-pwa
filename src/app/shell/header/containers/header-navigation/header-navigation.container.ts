import { ChangeDetectionStrategy, Component, Input, OnInit } from '@angular/core';
import { Observable } from 'rxjs';

import { ShoppingFacade } from 'ish-core/facades/shopping.facade';
import { CategoryView } from 'ish-core/models/category-view/category-view.model';

@Component({
  selector: 'ish-header-navigation-container',
  templateUrl: './header-navigation.container.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class HeaderNavigationContainerComponent implements OnInit {
  @Input() view: 'auto' | 'small' | 'full' = 'auto';

  categories$: Observable<CategoryView[]>;

  constructor(private shoppingFacade: ShoppingFacade) {}

  ngOnInit() {
    this.categories$ = this.shoppingFacade.topLevelCategories$;
  }
}
