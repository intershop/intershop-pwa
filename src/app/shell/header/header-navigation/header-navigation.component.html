<ul class="navbar-nav main-navigation-list">
  <ng-container *ngTemplateOutlet="categoryTemplate"></ng-container>
</ul>
<!--- Category Template -->
<ng-template #categoryTemplate>
  <li
    *ngFor="let category of categories$ | async"
    #submenu
    class="dropdown"
    [ngClass]="{ open: isOpened(category.uniqueId) }"
    (mouseover)="subMenuShow(submenu)"
    (mouseleave)="subMenuHide(submenu)"
    (click)="subMenuHide(submenu)"
  >
    <a
      [routerLink]="category.url"
      [attr.data-testing-id]="category.uniqueId + '-link'"
      [ngStyle]="{ width: !category.hasChildren ? '100%' : '' }"
    >
      {{ category.name }}
    </a>
    <ng-container *ngIf="category.hasChildren">
      <a class="dropdown-toggle" (click)="toggleOpen(category.uniqueId)">
        <fa-icon *ngIf="isOpened(category.uniqueId); else closed" [icon]="['fas', 'minus']"></fa-icon>
        <ng-template #closed> <fa-icon [icon]="['fas', 'plus']"></fa-icon> </ng-template>
      </a>
      <ish-sub-category-navigation
        [categoryUniqueId]="category.uniqueId"
        [subCategoriesDepth]="1"
        [view]="view"
      ></ish-sub-category-navigation>
    </ng-container>
  </li>
</ng-template>
