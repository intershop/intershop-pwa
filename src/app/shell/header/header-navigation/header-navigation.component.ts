import { ChangeDetectionStrategy, Component, Input, OnInit } from '@angular/core';
import { Observable } from 'rxjs';

import { ShoppingFacade } from 'ish-core/facades/shopping.facade';
import { NavigationCategory } from 'ish-core/models/navigation-category/navigation-category.model';

@Component({
  selector: 'ish-header-navigation',
  templateUrl: './header-navigation.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class HeaderNavigationComponent implements OnInit {
  @Input() view: 'auto' | 'small' | 'full' = 'auto';

  categories$: Observable<NavigationCategory[]>;

  private openedCategories: string[] = [];

  constructor(private shoppingFacade: ShoppingFacade) {}

  ngOnInit() {
    this.categories$ = this.shoppingFacade.navigationCategories$();
  }

  /**
   * Handle sub menu show.
   * Adds hover class to rendered element.
   * @param submenu The rendered sub menu element.
   */
  subMenuShow(submenu: HTMLElement) {
    submenu.classList.add('hover');
  }

  /**
   * Handle sub menu hide.
   * Removes hover class from rendered element.
   * @param submenu The rendered sub menu element.
   */
  subMenuHide(submenu: HTMLElement) {
    submenu.classList.remove('hover');
  }

  /**
   * Indicate if specific category is expanded.
   * @param category The category item.
   */
  isOpened(uniqueId: string): boolean {
    return this.openedCategories.includes(uniqueId);
  }

  /**
   * Toggle category open state.
   * @param category The category item.
   */
  toggleOpen(uniqueId: string) {
    const index = this.openedCategories.findIndex(id => id === uniqueId);
    index > -1 ? this.openedCategories.splice(index, 1) : this.openedCategories.push(uniqueId);
  }
}
