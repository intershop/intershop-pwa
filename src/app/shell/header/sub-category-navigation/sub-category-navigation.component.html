<ul
  *ngIf="navigationCategories$ | async as categories"
  class="category-level{{ subCategoriesDepth }}"
  [ngClass]="{ 'dropdown-menu': subCategoriesDepth === 1 }"
>
  <li
    *ngFor="let subcategory of categories"
    class="main-navigation-level{{ subCategoriesDepth }}-item"
    [ngClass]="{ open: isOpened(subcategory.uniqueId) }"
  >
    <a [routerLink]="subcategory.url" [ngStyle]="{ width: !subcategory.hasChildren ? '100%' : '' }">{{
      subcategory.name
    }}</a>
    <a
      *ngIf="subcategory.hasChildren"
      class="dropdown-toggle"
      [ngClass]="{ 'd-none': view === 'full' }"
      (click)="toggleOpen(subcategory.uniqueId)"
    >
      <fa-icon *ngIf="isOpened(subcategory.uniqueId); else closed" [icon]="['fas', 'minus']"></fa-icon>
      <ng-template #closed> <fa-icon [icon]="['fas', 'plus']"></fa-icon> </ng-template>
    </a>
    <ish-sub-category-navigation
      *ngIf="subCategoriesDepth < mainNavigationMaxSubCategoriesDepth && subcategory.hasChildren"
      [categoryUniqueId]="subcategory.uniqueId"
      [subCategoriesDepth]="subCategoriesDepth + 1"
    ></ish-sub-category-navigation>
  </li>
</ul>
