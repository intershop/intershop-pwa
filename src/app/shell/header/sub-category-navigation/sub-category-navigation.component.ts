import { ChangeDetectionStrategy, Component, Inject, Input, OnInit } from '@angular/core';
import { Observable } from 'rxjs';

import { MAIN_NAVIGATION_MAX_SUB_CATEGORIES_DEPTH } from 'ish-core/configurations/injection-keys';
import { ShoppingFacade } from 'ish-core/facades/shopping.facade';
import { NavigationCategory } from 'ish-core/models/navigation-category/navigation-category.model';

/**
 * The Sub Category Navigation Component displays second level category navigation.
 */
@Component({
  selector: 'ish-sub-category-navigation',
  templateUrl: './sub-category-navigation.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SubCategoryNavigationComponent implements OnInit {
  @Input() view = 'auto';
  @Input() categoryUniqueId: string;
  @Input() subCategoriesDepth: number;

  openedCategories: string[] = [];

  navigationCategories$: Observable<NavigationCategory[]>;

  constructor(
    private shoppingFacade: ShoppingFacade,
    @Inject(MAIN_NAVIGATION_MAX_SUB_CATEGORIES_DEPTH) public mainNavigationMaxSubCategoriesDepth: number
  ) {}

  ngOnInit() {
    this.navigationCategories$ = this.shoppingFacade.navigationCategories$(this.categoryUniqueId);
  }

  /**
   * Indicate if specific category is expanded.
   */
  isOpened(uniqueId: string): boolean {
    return this.openedCategories.includes(uniqueId);
  }

  /**
   * Toggle category open state.
   */
  toggleOpen(uniqueId: string) {
    const index = this.openedCategories.findIndex(id => id === uniqueId);
    index > -1 ? this.openedCategories.splice(index, 1) : this.openedCategories.push(uniqueId);
  }
}
