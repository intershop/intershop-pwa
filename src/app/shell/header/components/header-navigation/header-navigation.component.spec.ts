import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { FaIconComponent } from '@fortawesome/angular-fontawesome';
import { MockComponent } from 'ng-mocks';

import { createCategoryView } from 'ish-core/models/category-view/category-view.model';
import { Category } from 'ish-core/models/category/category.model';
import { CategoryRoutePipe } from 'ish-core/pipes/category-route.pipe';
import { categoryTree } from 'ish-core/utils/dev/test-data-utils';
import { SubCategoryNavigationComponent } from 'ish-shell/header/components/sub-category-navigation/sub-category-navigation.component';

import { HeaderNavigationComponent } from './header-navigation.component';

describe('Header Navigation Component', () => {
  let fixture: ComponentFixture<HeaderNavigationComponent>;
  let component: HeaderNavigationComponent;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [
        CategoryRoutePipe,
        HeaderNavigationComponent,
        MockComponent(FaIconComponent),
        MockComponent(SubCategoryNavigationComponent),
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(HeaderNavigationComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;

    const categories = categoryTree([
      { uniqueId: 'A', name: 'CAT_A', categoryPath: ['A'] },
      { uniqueId: 'B', name: 'CAT_B', categoryPath: ['B'] },
      { uniqueId: 'C', name: 'CAT_C', categoryPath: ['C'] },
    ] as Category[]);

    component.categories = [
      createCategoryView(categories, 'A'),
      createCategoryView(categories, 'B'),
      createCategoryView(categories, 'C'),
    ];
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchInlineSnapshot(`
      <ul class="navbar-nav main-navigation-list">
        <li class="dropdown">
          <a
            style="width: 100%;"
            ng-reflect-router-link="/category/A"
            data-testing-id="A-link"
            href="/category/A"
          >
            CAT_A </a
          ><ish-sub-category-navigation
            ng-reflect-view="auto"
            ng-reflect-sub-categories-depth="1"
          ></ish-sub-category-navigation>
        </li>
        <li class="dropdown">
          <a
            style="width: 100%;"
            ng-reflect-router-link="/category/B"
            data-testing-id="B-link"
            href="/category/B"
          >
            CAT_B </a
          ><ish-sub-category-navigation
            ng-reflect-view="auto"
            ng-reflect-sub-categories-depth="1"
          ></ish-sub-category-navigation>
        </li>
        <li class="dropdown">
          <a
            style="width: 100%;"
            ng-reflect-router-link="/category/C"
            data-testing-id="C-link"
            href="/category/C"
          >
            CAT_C </a
          ><ish-sub-category-navigation
            ng-reflect-view="auto"
            ng-reflect-sub-categories-depth="1"
          ></ish-sub-category-navigation>
        </li>
      </ul>
    `);
  });
});
