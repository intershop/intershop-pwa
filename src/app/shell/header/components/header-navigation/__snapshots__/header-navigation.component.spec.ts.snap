// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Navigation Component should be created 1`] = `
<ul
  class="navbar-nav main-navigation-list d-md-none accordion"
  ng-reflect-klass="navbar-nav main-navigation-lis"
  ng-reflect-ng-class="[object Object]"
>
  <li class="dropdown" ng-reflect-klass="dropdown" ng-reflect-ng-class="[object Object]">
    <a style="width: 100%;" ng-reflect-ng-style="[object Object]" data-testing-id="A-link">
      CAT_A </a
    ><ish-sub-category-navigation></ish-sub-category-navigation>
  </li>
  <li class="dropdown" ng-reflect-klass="dropdown" ng-reflect-ng-class="[object Object]">
    <a style="width: 100%;" ng-reflect-ng-style="[object Object]" data-testing-id="B-link">
      CAT_B </a
    ><ish-sub-category-navigation></ish-sub-category-navigation>
  </li>
  <li class="dropdown" ng-reflect-klass="dropdown" ng-reflect-ng-class="[object Object]">
    <a style="width: 100%;" ng-reflect-ng-style="[object Object]" data-testing-id="C-link">
      CAT_C </a
    ><ish-sub-category-navigation></ish-sub-category-navigation>
  </li>
</ul>
<ul
  class="navbar-nav main-navigation-list d-none d-md-block"
  ng-reflect-klass="navbar-nav main-navigation-lis"
  ng-reflect-ng-class="[object Object]"
>
  <li class="dropdown" ng-reflect-klass="dropdown" ng-reflect-ng-class="[object Object]">
    <a style="width: 100%;" ng-reflect-ng-style="[object Object]" data-testing-id="A-link">
      CAT_A </a
    ><ish-sub-category-navigation></ish-sub-category-navigation>
  </li>
  <li class="dropdown" ng-reflect-klass="dropdown" ng-reflect-ng-class="[object Object]">
    <a style="width: 100%;" ng-reflect-ng-style="[object Object]" data-testing-id="B-link">
      CAT_B </a
    ><ish-sub-category-navigation></ish-sub-category-navigation>
  </li>
  <li class="dropdown" ng-reflect-klass="dropdown" ng-reflect-ng-class="[object Object]">
    <a style="width: 100%;" ng-reflect-ng-style="[object Object]" data-testing-id="C-link">
      CAT_C </a
    ><ish-sub-category-navigation></ish-sub-category-navigation>
  </li>
</ul>
`;
